minorminer/utils/zephyr/node_edge.py

```diff
@@ -0,0 +1,819 @@
+# Copyright 2025 D-Wave
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+# ================================================================================================
+
+
+from __future__ import annotations
+
+from collections import namedtuple
+from enum import Enum
+from itertools import product
+from typing import Callable, Generator, Iterable
+
+from minorminer.utils.zephyr.coordinate_systems import (CartesianCoord, ZephyrCoord,
+                                                        cartesian_to_zephyr, zephyr_to_cartesian)
+from minorminer.utils.zephyr.plane_shift import ZPlaneShift
+ZShape = namedtuple("ZShape", ["m", "t"], defaults=(None, None))
+
+
+
+class EdgeKind(Enum):
+    INTERNAL = 1
+    EXTERNAL = 2
+    ODD = 3
+
+
+class NodeKind(Enum):
+    VERTICAL = 0
+    HORIZONTAL = 1
+
+
+
+class Edge:
+    """Initializes an Edge with nodes x, y.
+
+    Args:
+        x : One endpoint of edge.
+        y : Another endpoint of edge.
+    """
+
+    def __init__(
+        self,
+        x,
+        y,
+    ) -> None:
+        self._edge = self._set_edge(x, y)
+
+    def _set_edge(self, x, y):
+        """Reutrns ordered tuple corresponding to the set {x, y}."""
+        if x < y:
+            return (x, y)
+        else:
+            return (y, x)
+
+    def __hash__(self):
+        return hash(self._edge)
+
+    def __getitem__(self, index: int) -> int:
+        return self._edge[index]
+
+    def __eq__(self, other: Edge):
+        return self._edge == other._edge
+
+    def __str__(self) -> str:
+        return f"{type(self).__name__}{self._edge}"
+
+    def __repr__(self) -> str:
+        return f"{type(self).__name__}{self._edge}"
+
+
+class ZEdge(Edge):
+    """Initializes a ZEdge with 'ZNode' nodes x, y.
+
+    Args:
+        x (ZNode): One endpoint of edge.
+        y (ZNode): Another endpoint of edge.
```

Comment:
A few of my older comments seem to have been missed. Readding them.
```suggestion
        x (ZNode): Endpoint of edge. Must have same shape as ``y``.
        y (ZNode): Endpoint of edge. Must have same shape as ``x``
```